In [6]:
from sklearn.datasets import load_iris

In [7]:
iris=load_iris()

In [8]:
X=iris.data
Y=iris.target

In [9]:
feature_names=iris.feature_names
target_names=iris.target_names

In [10]:
feature_names


['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [11]:
target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [12]:
type(X)

numpy.ndarray

In [13]:
X[:5]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2]])

In [14]:
X.shape

(150, 4)

In [15]:
import pandas as pd
data=pd.read_csv("C:\Python36-32\weather.csv")

In [16]:
data

Outlook Temperature Humidity  Windy Play
0   overcast         hot     high  False  yes
1   overcast        cool   normal   True  yes
2   overcast        mild     high   True  yes
3   overcast         hot   normal  False  yes
4      rainy        mild     high  False  yes
5      rainy        cool   normal  False  yes
6      rainy        cool   normal   True   no
7      rainy        mild   normal  False  yes
8      rainy        mild     high   True   no
9      sunny         hot     high  False   no
10     sunny         hot     high   True   no
11     sunny        mild     high  False   no
12     sunny        cool   normal  False  yes
13     sunny        mild   normal   True  yes

In [17]:
data.shape

(14, 5)

In [18]:
data.columns

Index(['Outlook', 'Temperature', 'Humidity', 'Windy', 'Play'], dtype='object')

In [19]:
z=data[data.columns[:-1]]

In [20]:
z

Outlook Temperature Humidity  Windy
0   overcast         hot     high  False
1   overcast        cool   normal   True
2   overcast        mild     high   True
3   overcast         hot   normal  False
4      rainy        mild     high  False
5      rainy        cool   normal  False
6      rainy        cool   normal   True
7      rainy        mild   normal  False
8      rainy        mild     high   True
9      sunny         hot     high  False
10     sunny         hot     high   True
11     sunny        mild     high  False
12     sunny        cool   normal  False
13     sunny        mild   normal   True

In [21]:
v=data[data.columns[-1]]

In [22]:
v

0     yes
1     yes
2     yes
3     yes
4     yes
5     yes
6      no
7     yes
8      no
9      no
10     no
11     no
12    yes
13    yes
Name: Play, dtype: object

In [23]:
from sklearn.model_selection import train_test_split

In [24]:
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=.4,random_state=1)

In [25]:
X_train.shape

(90, 4)

In [26]:
Y_train.shape

(90,)

In [27]:
X_test.shape

(60, 4)

In [28]:
Y_test.shape

(60,)

In [29]:
Y_train

array([0, 2, 2, 0, 0, 2, 0, 2, 2, 1, 1, 2, 2, 0, 1, 1, 2, 1, 2, 1, 0, 0,
       0, 2, 0, 1, 2, 2, 0, 0, 1, 0, 2, 1, 2, 2, 1, 2, 2, 1, 0, 1, 0, 1,
       1, 0, 1, 0, 0, 2, 2, 2, 0, 0, 1, 0, 2, 0, 2, 2, 0, 2, 0, 1, 0, 1,
       1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 2, 0, 0, 2, 1, 2, 1, 2, 2, 1,
       2, 0])

test_size: It is the ratio of test data to the given data. For example, 
    setting test_size = 0.4 for 150 rows of X produces test data of 150 x 0.4 = 60 rows.
random_state: If you use random_state = some_number, then you can guarantee that your split will be always the same. 
    This is useful if you want reproducible results, for example in testing for consistency 
    in the documentation (so that everybody can see the same numbers).

In [30]:
from sklearn.neighbors import KNeighborsClassifier

In [31]:
knn=KNeighborsClassifier(n_neighbors=3)

In [32]:
knn

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform')

In [33]:
knn.fit(X_train,Y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform')

In [34]:
y_pred=knn.predict(X_test)

In [35]:
knn

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform')

In [36]:
y_pred

array([0, 1, 1, 0, 2, 1, 2, 0, 0, 2, 1, 0, 2, 1, 1, 0, 1, 1, 0, 0, 1, 1,
       1, 0, 2, 1, 0, 0, 1, 2, 1, 2, 1, 2, 2, 0, 1, 0, 1, 2, 2, 0, 1, 2,
       1, 2, 0, 0, 0, 1, 0, 0, 2, 2, 2, 2, 2, 1, 2, 1])

In [37]:
y_pred.shape

(60,)

In [38]:
from sklearn import metrics

In [39]:
print("knn model accuracy:",metrics.accuracy_score(Y_test,y_pred))

knn model accuracy: 0.9833333333333333


In [40]:
sample = [[1.3,2.5, 4.2, 2.5], [2.5, 3.4, 1.5, 4.6]]

In [41]:
preds = knn.predict(sample)

In [42]:
from sklearn.externals import joblib

In [43]:
joblib.dump(knn,'iris_knn.pkl')

['iris_knn.pkl']

if you are not interested in training your classifier again and again and use the pre-trained classifier,
one can save their classifier using joblib. All you need to do is:

joblib.dump(knn, 'iris_knn.pkl')


In case you want to load an already saved classifier, use the following method:

knn = joblib.load('iris_knn.pkl')

# As we approach to end of this article, here are some benefits of using 
scikit-learn over some other machine learning libraries(like R):

    Consistent interface to machine learning models
    Provides many tuning parameters but with sensible defaults
    Exceptional documentation
    Rich set of functionality for companion tasks.
    Active community for development and support.
